File: design-pattern-python/01_builder.py
class Computer:
    cpu = "cpu - not defined"
    ram = "ram - not defined"
    storage = "storage - not defined"
    gpu: str | None = None

    def display_configuration(self):
        print(f"Configuration de la computadora\n"
              f"CPU: {self.cpu}\n"
              f"RAM: {self.ram}\n"
              f"Almacenamiento: {self.storage}\n"
              f"GPU: {"Sin gpu" if self.gpu is None else self.gpu}")


class ComputerBuilder:
    def __init__(self):
        self.computer = Computer()

    def set_cpu(self, cpu):
        self.computer.cpu = cpu
        return self

    def set_ram(self, ram):
        self.computer.ram = ram
        return self

    def set_storage(self, storage):
        self.computer.storage = storage
        return self

    def set_gpu(self, gpu):
        self.computer.gpu = gpu
        return self

    def build(self):
        return self.computer


def main():
    basic_computer = ComputerBuilder().set_cpu("Intel core 2 duo").set_ram("4GB").set_storage("256GB").build()

    print("Computadora básica:")
    basic_computer.display_configuration()

    gamer_computer = ComputerBuilder().set_cpu("i7 14700k").set_cpu("i9 14900k").set_ram("128GB").set_storage(
        "4TB").set_gpu("RTX 5090 ti").build()

    print("Computadora gamer:")
    gamer_computer.display_configuration()


main()

File: design-pattern-javascript/01-creacionales/01-builder.js
/**
 * ! Patrón Builder:
 * Es un patrón de diseño creacional que nos permite construir objetos complejos
 * paso a paso.
 *
 * El patrón nos permite producir distintos tipos y representaciones
 * de un objeto empleando el mismo código de construcción.
 *
 * * Es útil cuando necesitamos construir un objeto complejo con muchas partes
 * * y queremos que el proceso de construcción sea independiente de las partes
 * * que lo componen.
 *
 * https://refactoring.guru/es/design-patterns/builder
 */
import { COLORS } from "../helpers/colors.js";

class Computer {
	cpu = "cpu - not defined";
	ram = "ram - not defined";
	storage = "storage - not defined";
	gpu;

	displayConfiguration() {
		console.log(`Configuración de la computadora
      CPU: ${this.cpu}
      RAM: ${this.ram}
      Almacenamiento: ${this.storage}
      GPU: ${this.gpu ?? "Sin gpu"}
      `);
	}
}

class ComputerBuilder {
	computer;

	constructor() {
		this.computer = new Computer();
	}

	setCpu(cpu) {
		this.computer.cpu = cpu;
		return this;
	}

	setRam(ram) {
		this.computer.ram = ram;
		return this;
	}

	setStorage(storage) {
		this.computer.storage = storage;
		return this;
	}

	setGpu(gpu) {
		this.computer.gpu = gpu;
		return this;
	}

	build() {
		return this.computer;
	}
}

function main() {
	const basicComputer = new ComputerBuilder()
		.setCpu("Intel Core 2 Duo")
		.setRam("4GB")
		.setStorage("256GB")
		.build();

	console.log("%cComputadora básica:", COLORS.blue);
	basicComputer.displayConfiguration();

	const gamerComputer = new ComputerBuilder()
		.setCpu("i7 14700k")
		.setCpu("i9 14900k")
		.setRam("128GB")
		.setStorage("4TB")
		.setGpu("RTX 5090 Ti")
		.build();

	console.log("%cComputadora gamer:", COLORS.red);
	gamerComputer.displayConfiguration();
}

main();
